pyadjoint/checkpointing.py

```diff
@@ -0,0 +1,305 @@
+
+    Attributes
+    ----------
+    steps : int
+        The number of forward steps in the initial forward calculation.
+    """
+    def __init__(self, cp_action=(), steps=0):
+        # Steps is only intended to provide a rough indicator for progress
+        # bars.
+        super().__init__(cp_action)
+        self.steps = steps
+
+
+def process_schedule(schedule):
+    """Process a checkpoint schedule into forward and adjoint schedules.
+
+    Parameters
+    ----------
+    schedule : checkpoint_schedules.schedule
+        A schedule provided by the checkpoint_schedules package.
+
+    Returns
+    -------
+    list, list
+        Forward and adjoint list of schedules.
+    """
+    schedule = list(schedule)
+    index = 0
+    forward_steps = 0
+
+    while not isinstance(schedule[index], EndForward):
+        if isinstance(schedule[index], Forward):
+            forward_steps += len(schedule[index])
+        index += 1
+    end_forward = index + 1
+    reverse_steps = 0
+
+    while not isinstance(schedule[index], EndReverse):
+        if isinstance(schedule[index], Reverse) or isinstance(schedule[index], Forward):  # noqa: E501
+            reverse_steps += len(schedule[index])
+        index += 1
+    forward = AdjointSchedule(schedule[:end_forward], forward_steps)
+    reverse = AdjointSchedule(schedule[end_forward:], reverse_steps)
+
+    return forward, reverse
+
+
+class CheckpointManager:
+    """Manage the executions of the forward and adjoint solvers.
+
+    Attributes
+    ----------
+    schedule : checkpoint_schedules.schedule
+        A schedule provided by the checkpoint_schedules package.
+    tape : Tape
+        A list of blocks :class:`Block` instances.
+        Each block represents one operation in the forward model.
+    """
+    def __init__(self, schedule, tape):
+        if not isinstance(schedule, Revolve):
+            raise CheckpointError("Only Revolve schedules are supported.")
+
+        self.fwd_schedule, self.rev_schedule = process_schedule(schedule)
+        if schedule.uses_storage_type(StorageType.DISK) and not tape._package_data:  # noqa: E501
+            raise CheckpointError(
+                "The schedule employs disk checkpointing but it is not configured."  # noqa: E501
+            )
+        self.tape = tape
+        self._schedule = schedule
+        self.timesteps = schedule.max_n
+        self.mode = Mode.RECORD
+        self._iterator = iter(self.fwd_schedule)
+        self._current = next(self._iterator)
+        self._configuration_step = None
+        self._stored_timesteps = []
+        self._shedule = schedule
+        self._checkpointable_state = set()
+        # Tell the tape to only checkpoint input data until told otherwise.
+        self.tape.latest_checkpoint = 0
+        # Process any initial instructions on the tape.
+        self.end_timestep(-1)
+
+    def end_timestep(self, timestep):
+        """Process any initial instructions on the tape.
```

Comment:
I don't think that this docstring really matches the name of the function. It's not very clear what happens.